kernel/src/process/clone.rs

```diff
@@ -541,14 +591,47 @@ fn create_child_process(
     child_proc
 }
 
-fn set_parent_and_group(parent: &Process, child: &Arc<Process>) {
-    // Lock order: process table -> children -> group of process
-    // -> group inner -> session inner
-    let mut process_table_mut = process_table::process_table_mut();
+fn clone_pid_namespace(parent: &Process, clone_flags: CloneFlags) -> Result<Arc<PidNamespace>> {
+    let parent_pid_ns = parent.pid_namespace();
+    let pid_ns_for_children = parent.pid_ns_for_children().get();
+
+    let child_pid_ns = if clone_flags.contains(CloneFlags::CLONE_NEWPID) {
+        if pid_ns_for_children.is_some() {
+            return_errno_with_message!(
+                Errno::EINVAL,
+                "cannot use CLONE_NEWPID when `pid_ns_for_children` is set"
+            );
+        }
+        PidNamespace::new_child(parent_pid_ns)?
+    } else if let Some(pid_ns_for_children) = pid_ns_for_children {
+        pid_ns_for_children.clone()
+    } else {
+        parent_pid_ns.clone()
+    };
 
+    if child_pid_ns.is_init_proc_terminated() {
```

Comment:
This is racy. We can create a new process in a dead process namespace due to race conditions, and the new process will not receive the SIGKILL signal due to race conditions.

Reply:
Oh, thanks for your reminder. We may need to recheck the status before the child returns to user space or held some init process lock here.